File: ESTRUCTURA DE DATOS/DEBERES/SEGUNDO_PARCIAL/DEBER2/PROYECTO_NRC1978/LISTA_SIMPLE_PYTHON/Ordenamiento.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include "Coche.h"
#include <functional>

template <typename T, typename Comparator>
int partition(std::vector<T>& lista, int low, int high, Comparator comp) {
    T pivot = lista[high];
    int i = low - 1;
    for (int j = low; j < high; j++) {
        if (comp(lista[j], pivot)) {
            i++;
            std::swap(lista[i], lista[j]);
        }
    }
    std::swap(lista[i + 1], lista[high]);
    return i + 1;
}

template <typename T, typename Comparator>
void quickSort(std::vector<T>& lista, int low, int high, Comparator comp) {
    if (low < high) {
        int pi = partition(lista, low, high, comp);
        quickSort(lista, low, pi - 1, comp);
        quickSort(lista, pi + 1, high, comp);
    }
}

template <typename T, typename Comparator>
void bubbleSortPass(std::vector<T>& lista, Comparator comp, int n) {
    // Función auxiliar que realiza una pasada completa de intercambios
    if (n <= 1) return;

    for (int j = 0; j < n - 1; j++) {
        if (!comp(lista[j], lista[j + 1])) {
            std::swap(lista[j], lista[j + 1]);
        }
    }
}

template <typename T, typename Comparator>
void bubbleSortRecursive(std::vector<T>& lista, Comparator comp, int n) {
    // Función recursiva principal
    if (n <= 1) return;
    // Realizar una pasada completa de intercambios
    bubbleSortPass(lista, comp, n);
    // Llamada recursiva con un rango reducido
    bubbleSortRecursive(lista, comp, n - 1);
}

// Función wrapper para iniciar el ordenamiento
template <typename T, typename Comparator>
void bubbleSort(std::vector<T>& lista, Comparator comp) {
    bubbleSortRecursive(lista, comp, lista.size());
}

template <typename T, typename Comparator>
void ordenarLista(ListaSimple<T>& lista, Comparator comp) {
    std::vector<T> elementos;

    // Extraer los elementos de la lista simplemente enlazada
    Nodo<T>* aux = lista.getPrimero();
    while (aux != nullptr) {
        elementos.push_back(aux->getDato());
        aux = aux->getSiguiente();
    }

    // Ordenar los elementos usando QuickSort
    quickSort(elementos, 0, elementos.size() - 1, comp);

    // Volver a insertar los elementos ordenados en la lista
    aux = lista.getPrimero();
    int index = 0;
    while (aux != nullptr) {
        aux->setDato(elementos[index++]);
        aux = aux->getSiguiente();
    }

    // Mostrar la lista ordenada
    lista.mostrar(lista.getPrimero());
}

template <typename T, typename Comparator>
void ordenarListaBubbleSort(ListaSimple<T>& lista, Comparator comp) {
    std::vector<T> elementos;

    // Extraer los elementos de la lista
    Nodo<T>* aux = lista.getPrimero();
    while (aux != nullptr) {
        elementos.push_back(aux->getDato());
        aux = aux->getSiguiente();
    }

    // Ordenar usando Bubble Sort
    bubbleSort(elementos, comp);

    // Volver a insertar los elementos ordenados en la lista
    aux = lista.getPrimero();
    int index = 0;
    while (aux != nullptr) {
        aux->setDato(elementos[index++]);
        aux = aux->getSiguiente();
    }

    lista.mostrar(lista.getPrimero());
}

template <typename T>
void ordenarListaBucket(ListaSimple<T>& lista, std::function<std::string(const T&)> getKey) {
    std::vector<T> elementos;

    // Extraer los elementos de la lista simplemente enlazada
    Nodo<T>* aux = lista.getPrimero();
    while (aux != nullptr) {
        elementos.push_back(aux->getDato());
        aux = aux->getSiguiente();
    }

    // Determinar el número de buckets (256 para cubrir caracteres ASCII)
    const int bucketCount = 256;
    std::vector<std::vector<T>> buckets(bucketCount);

    // Distribuir los elementos en los buckets
    for (const auto& item : elementos) {
        std::string key = getKey(item);
        if (!key.empty()) {
            char firstChar = tolower(key[0]); // Usar el primer carácter de la clave
            int bucketIndex = static_cast<int>(firstChar);
            buckets[bucketIndex].push_back(item);
        }
    }

    // Limpiar el vector de elementos
    elementos.clear();

    // Ordenar cada bucket y volver a insertar en el vector de elementos
    for (auto& bucket : buckets) {
        std::sort(bucket.begin(), bucket.end(), [&getKey](const T& a, const T& b) {
            return getKey(a) < getKey(b); // Usar la clave para ordenar
        });
        elementos.insert(elementos.end(), bucket.begin(), bucket.end());
    }

    // Volver a insertar los elementos ordenados en la lista simplemente enlazada
    aux = lista.getPrimero();
    int index = 0;
    while (aux != nullptr) {
        aux->setDato(elementos[index++]);
        aux = aux->getSiguiente();
    }

    // Mostrar la lista ordenada
    lista.mostrar(lista.getPrimero());
}

template <typename T, typename Comparator>
void ordenarListaShellSort(ListaSimple<T>& lista, Comparator comp) {
    std::vector<T> elementos;

    // Extraer los elementos de la lista
    Nodo<T>* aux = lista.getPrimero();
    while (aux != nullptr) {
        elementos.push_back(aux->getDato());
        aux = aux->getSiguiente();
    }

    // Ordenar usando Shell Sort
    auto shellSort = [&](std::vector<T>& elementos) {
        int n = elementos.size();
        for (int gap = n / 2; gap > 0; gap /= 2) {
            for (int i = gap; i < n; i++) {
                T temp = elementos[i];
                int j = i;
                while (j >= gap && comp(temp, elementos[j - gap])) {
                    elementos[j] = elementos[j - gap];
                    j -= gap;
                }
                elementos[j] = temp;
            }
        }
    };

    shellSort(elementos);

    // Volver a insertar los elementos ordenados
    aux = lista.getPrimero();
    int index = 0;
    while (aux != nullptr) {
        aux->setDato(elementos[index++]);
        aux = aux->getSiguiente();
    }

    lista.mostrar(lista.getPrimero());
}

template <typename T, typename KeyExtractor>
void radixSortByString(std::vector<T>& elementos, KeyExtractor getKey) {
    int maxLength = 0;
    for (const auto& item : elementos) {
        int length = static_cast<int>(getKey(item).length());
        maxLength = std::max(maxLength, length);
    }

    for (int i = maxLength - 1; i >= 0; i--) {
        std::vector<std::vector<T>> buckets(256);
        for (const auto& item : elementos) {
            std::string key = getKey(item);
            int charIndex = static_cast<int>(key.length()) - 1 - i;
            int charValue = (charIndex >= 0) ? static_cast<int>(key[charIndex]) : 0;
            buckets[charValue].push_back(item);
        }

        elementos.clear();
        for (auto& bucket : buckets) {
            for (const auto& item : bucket) {
                elementos.push_back(item);
            }
        }
    }
}

template <typename T, typename KeyExtractor>
void radixSortByInt(std::vector<T>& elementos, KeyExtractor getKey) {
    int maxValue = 0;
    for (const auto& item : elementos) {
        int value = getKey(item);
        maxValue = std::max(maxValue, value);
    }

    for (int exp = 1; maxValue / exp > 0; exp *= 10) {
        std::vector<std::vector<T>> buckets(10);
        for (const auto& item : elementos) {
            int digit = (getKey(item) / exp) % 10;
            buckets[digit].push_back(item);
        }

        elementos.clear();
        for (auto& bucket : buckets) {
            for (const auto& item : bucket) {
                elementos.push_back(item);
            }
        }
    }
}


template <typename T, typename KeyExtractor>
void ordenarListaPorRadix(ListaSimple<T>& lista, KeyExtractor getKey) {
    std::vector<T> elementos;

    // Extraer los elementos de la lista
    Nodo<T>* aux = lista.getPrimero();
    while (aux != nullptr) {
        elementos.push_back(aux->getDato());
        aux = aux->getSiguiente();
    }

    // Determinar si se ordena por cadena o entero
    using KeyType = typename std::invoke_result<KeyExtractor, T>::type;
    if constexpr (std::is_same<KeyType, std::string>::value) {
        radixSortByString(elementos, getKey);
    } else if constexpr (std::is_integral<KeyType>::value) {
        radixSortByInt(elementos, getKey);
    } else {
        static_assert(false, "El tipo de clave no es compatible con Radix Sort");
    }

    // Volver a insertar los elementos ordenados
    aux = lista.getPrimero();
    int index = 0;
    while (aux != nullptr) {
        aux->setDato(elementos[index++]);
        aux = aux->getSiguiente();
    }

    lista.mostrar(lista.getPrimero());
}

File: ESTRUCTURA DE DATOS/DEBERES/SEGUNDO_PARCIAL/DEBER2/PROYECTO_NRC1978/LISTA_SIMPLE_PYTHON/PYTHON/ordenamiento.py
from typing import List, Callable
from nodo import Nodo  # Para la clase Nodo
from lista import ListaSimple  # Para la clase ListaSimple

class Ordenamiento:
    # QuickSort
    @staticmethod
    def partition(lista: List, low: int, high: int, comp: Callable):
        pivot = lista[high]
        i = low - 1
        for j in range(low, high):
            if comp(lista[j], pivot):
                i += 1
                lista[i], lista[j] = lista[j], lista[i]
        lista[i + 1], lista[high] = lista[high], lista[i + 1]
        return i + 1

    @staticmethod
    def quick_sort(lista: List, low: int, high: int, comp: Callable):
        if low < high:
            pi = Ordenamiento.partition(lista, low, high, comp)
            Ordenamiento.quick_sort(lista, low, pi - 1, comp)
            Ordenamiento.quick_sort(lista, pi + 1, high, comp)

    # BubbleSort
    @staticmethod
    def bubble_sort_pass(lista: List, comp: Callable, n: int):
        if n <= 1:
            return
        for j in range(n - 1):
            if not comp(lista[j], lista[j + 1]):
                lista[j], lista[j + 1] = lista[j + 1], lista[j]

    @staticmethod
    def bubble_sort_recursive(lista: List, comp: Callable, n: int):
        if n <= 1:
            return
        Ordenamiento.bubble_sort_pass(lista, comp, n)
        Ordenamiento.bubble_sort_recursive(lista, comp, n - 1)

    @staticmethod
    def bubble_sort(lista: List, comp: Callable):
        Ordenamiento.bubble_sort_recursive(lista, comp, len(lista))

    # ShellSort
    @staticmethod
    def ordenar_lista_shell_sort(lista: ListaSimple, comp: Callable):
        elementos = []
        aux = lista.get_primero()
        while aux:
            elementos.append(aux.get_dato())
            aux = aux.get_siguiente()

        def shell_sort(elementos: List):
            n = len(elementos)
            gap = n // 2
            while gap > 0:
                for i in range(gap, n):
                    temp = elementos[i]
                    j = i
                    while j >= gap and comp(temp, elementos[j - gap]):
                        elementos[j] = elementos[j - gap]
                        j -= gap
                    elementos[j] = temp
                gap //= 2

        shell_sort(elementos)

        aux = lista.get_primero()
        index = 0
        while aux:
            aux.set_dato(elementos[index])
            aux = aux.get_siguiente()
            index += 1

        lista.mostrar(lista.get_primero())

    # RadixSort
    @staticmethod
    def radix_sort_by_string(elementos: List, get_key: Callable):
        max_length = max(len(get_key(item)) for item in elementos)
        for i in range(max_length - 1, -1, -1):
            buckets = [[] for _ in range(256)]
            for item in elementos:
                key = get_key(item)
                char_index = len(key) - 1 - i
                char_value = ord(key[char_index]) if char_index >= 0 else 0
                buckets[char_value].append(item)

            elementos.clear()
            for bucket in buckets:
                for item in bucket:
                    elementos.append(item)

    @staticmethod
    def radix_sort_by_int(elementos: List, get_key: Callable):
        max_value = max(get_key(item) for item in elementos)
        exp = 1
        while max_value // exp > 0:
            buckets = [[] for _ in range(10)]
            for item in elementos:
                digit = (get_key(item) // exp) % 10
                buckets[digit].append(item)

            elementos.clear()
            for bucket in buckets:
                for item in bucket:
                    elementos.append(item)
            exp *= 10

    @staticmethod
    def ordenar_lista_por_radix(lista: ListaSimple, get_key: Callable):
        elementos = []
        aux = lista.get_primero()
        while aux:
            elementos.append(aux.get_dato())
            aux = aux.get_siguiente()

        key_type = type(get_key(elementos[0]))
        if isinstance(get_key(elementos[0]), str):
            Ordenamiento.radix_sort_by_string(elementos, get_key)
        elif isinstance(get_key(elementos[0]), int):
            Ordenamiento.radix_sort_by_int(elementos, get_key)
        else:
            raise TypeError("El tipo de clave no es compatible con Radix Sort")

        aux = lista.get_primero()
        index = 0
        while aux:
            aux.set_dato(elementos[index])
            aux = aux.get_siguiente()
            index += 1

        lista.mostrar(lista.get_primero())

    # Ordenar usando QuickSort
    @staticmethod
    def ordenar_lista(lista: ListaSimple, comp: Callable):
        elementos = []
        aux = lista.get_primero()
        while aux:
            elementos.append(aux.get_dato())
            aux = aux.get_siguiente()

        Ordenamiento.quick_sort(elementos, 0, len(elementos) - 1, comp)

        aux = lista.get_primero()
        index = 0
        while aux:
            aux.set_dato(elementos[index])
            aux = aux.get_siguiente()
            index += 1

        lista.mostrar(lista.get_primero())

    # Ordenar usando BubbleSort
    @staticmethod
    def ordenar_lista_bubble_sort(lista: ListaSimple, comp: Callable):
        elementos = []
        aux = lista.get_primero()
        while aux:
            elementos.append(aux.get_dato())
            aux = aux.get_siguiente()

        Ordenamiento.bubble_sort(elementos, comp)

        aux = lista.get_primero()
        index = 0
        while aux:
            aux.set_dato(elementos[index])
            aux = aux.get_siguiente()
            index += 1

        lista.mostrar(lista.get_primero())

    # Ordenar usando Bucket Sort
    @staticmethod
    def ordenar_lista_bucket(lista: ListaSimple, get_key: Callable):
        elementos = []
        aux = lista.get_primero()
        while aux:
            elementos.append(aux.get_dato())
            aux = aux.get_siguiente()

        bucket_count = 256
        buckets = [[] for _ in range(bucket_count)]

        for item in elementos:
            key = get_key(item)
            if key:
                first_char = key[0].lower()
                bucket_index = ord(first_char)
                buckets[bucket_index].append(item)

        elementos.clear()

        for bucket in buckets:
            bucket.sort(key=get_key)
            elementos.extend(bucket)

        aux = lista.get_primero()
        index = 0
        while aux:
            aux.set_dato(elementos[index])
            aux = aux.get_siguiente()
            index += 1

        lista.mostrar(lista.get_primero())
